manta-js/package/src/sdk.ts

```
@@ -0,0 +1,538 @@
+                networkType
+            );
+            const privateAddressBytes = [
+                ...privateAddressRaw.receiving_key
+            ];
+            const privateAddress = base58Encode(privateAddressBytes);
+            this.walletIsBusy = false;
+            return privateAddress;
+        } catch (e) {
+            this.walletIsBusy = false;
+            console.error("Failed to fetch ZkAddress.",e);
+        }
+    }
+
+    /// Performs full wallet recovery. Restarts `self` with an empty state and
+    /// performs a synchronization against the signer and ledger to catch up to
+    /// the current checkpoint and balance state.
+    ///
+    /// Requirements: Must be called once after creating an instance of MantaPrivateWallet
+    /// and must be called before walletSync().
+    async initalWalletSync(): Promise<void> {
+        try {
+            await this.waitForWallet();
+            this.walletIsBusy = true;
+            this.log('Beginning initial sync');
+            const networkType = this.wasm.Network.from_string(`"${this.network}"`);
+            const startTime = performance.now();
+            await this.wasmWallet.restart(networkType);
+            const endTime = performance.now();
+            this.log(`Initial sync finished in ${(endTime - startTime) / 1000} seconds`);
+            this.walletIsBusy = false;
+            this.initialSyncIsFinished = true;
+        } catch (e) {
+            this.walletIsBusy = false;
+            console.error("Initial sync failed.",e);
+        }
+    }
+
+    /// Pulls data from the ledger, synchronizing the currently connected wallet and
+    /// balance state. This method runs until all the ledger data has arrived at and
+    /// has been synchronized with the wallet.
+    async walletSync(): Promise<void> {
+        try {
+            if (!this.initialSyncIsFinished) {
+                console.error("Must call initalWalletSync before walletSync!");
```

We should throw an error rather than logging an error here. Bad bad things can probably happen if this rule isn't followed, and external devs may not know this rule, so we need to enforce it very strictly.